// ng-uikit-pro-standard/esm2015/lib/pro/accordion/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SqueezeBoxComponent } from './components/squeezebox';
import { SBItemComponent } from './components/sb-item';
import { SBItemHeadComponent } from './components/sb-item.head';
import { SBItemBodyComponent } from './components/sb-item.body';
import * as ɵngcc0 from '@angular/core';
export const SQUEEZEBOX_COMPONENTS = [SqueezeBoxComponent, SBItemComponent, SBItemHeadComponent, SBItemBodyComponent];
export { SBItemComponent } from './components/sb-item';
export { SBItemHeadComponent } from './components/sb-item.head';
export { SBItemBodyComponent } from './components/sb-item.body';
export { SqueezeBoxComponent } from './components/squeezebox';
export class AccordionModule {
}
AccordionModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AccordionModule });
AccordionModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AccordionModule_Factory(t) { return new (t || AccordionModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AccordionModule, { declarations: function () { return [SqueezeBoxComponent, SBItemComponent, SBItemHeadComponent, SBItemBodyComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [SqueezeBoxComponent, SBItemComponent, SBItemHeadComponent, SBItemBodyComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AccordionModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [SQUEEZEBOX_COMPONENTS],
                exports: [SQUEEZEBOX_COMPONENTS]
            }]
    }], null, null); })();
