// ng-uikit-pro-standard/esm2015/lib/pro/stepper/stepper.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MdbStepComponent } from './step.component';
import { MdbStepperComponent } from './stepper.component';
import * as ɵngcc0 from '@angular/core';
export class StepperModule {
}
StepperModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: StepperModule });
StepperModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function StepperModule_Factory(t) { return new (t || StepperModule)(); }, imports: [[
            CommonModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepperModule, { declarations: function () { return [MdbStepperComponent,
        MdbStepComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [MdbStepperComponent,
        MdbStepComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    MdbStepperComponent,
                    MdbStepComponent
                ],
                exports: [
                    MdbStepperComponent,
                    MdbStepComponent
                ],
                imports: [
                    CommonModule,
                ]
            }]
    }], null, null); })();
