// angular-calendar/esm5/modules/day/calendar-all-day-event.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../common/calendar-event-actions.component';
import * as ɵngcc3 from '../common/calendar-event-title.component';
import * as ɵngcc4 from '../common/click.directive';

function CalendarAllDayEventComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelement(1, "mwl-calendar-event-actions", 3);
    ɵngcc0.ɵɵtext(2, " ");
    ɵngcc0.ɵɵelementStart(3, "mwl-calendar-event-title", 4);
    ɵngcc0.ɵɵlistener("mwlClick", function CalendarAllDayEventComponent_ng_template_0_Template_mwl_calendar_event_title_mwlClick_3_listener() { var eventClicked_r4 = ctx.eventClicked; return eventClicked_r4.emit(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var event_r3 = ctx.event;
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("background-color", event_r3.color == null ? null : event_r3.color.secondary)("border-color", event_r3.color == null ? null : event_r3.color.primary);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("event", event_r3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("event", event_r3)("customTemplate", ctx_r1.eventTitleTemplate);
} }
function CalendarAllDayEventComponent_ng_template_2_Template(rf, ctx) { }
var _c0 = function (a0, a1) { return { event: a0, eventClicked: a1 }; };
var CalendarAllDayEventComponent = /** @class */ (function () {
    function CalendarAllDayEventComponent() {
        this.eventClicked = new EventEmitter();
    }
    /** @nocollapse */
    CalendarAllDayEventComponent.propDecorators = {
        "event": [{ type: Input },],
        "customTemplate": [{ type: Input },],
        "eventTitleTemplate": [{ type: Input },],
        "eventClicked": [{ type: Output },],
    };
CalendarAllDayEventComponent.ɵfac = function CalendarAllDayEventComponent_Factory(t) { return new (t || CalendarAllDayEventComponent)(); };
CalendarAllDayEventComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CalendarAllDayEventComponent, selectors: [["mwl-calendar-all-day-event"]], inputs: { event: "event", customTemplate: "customTemplate", eventTitleTemplate: "eventTitleTemplate" }, outputs: { eventClicked: "eventClicked" }, decls: 3, vars: 5, consts: [["defaultTemplate", ""], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "cal-all-day-event"], [3, "event"], ["view", "day", 3, "event", "customTemplate", "mwlClick"]], template: function CalendarAllDayEventComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, CalendarAllDayEventComponent_ng_template_0_Template, 4, 7, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(2, CalendarAllDayEventComponent_ng_template_2_Template, 0, 0, "ng-template", 1);
    } if (rf & 2) {
        var _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.customTemplate || _r0)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c0, ctx.event, ctx.eventClicked));
    } }, directives: [ɵngcc1.NgTemplateOutlet, ɵngcc2.CalendarEventActionsComponent, ɵngcc3.CalendarEventTitleComponent, ɵngcc4.ClickDirective], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CalendarAllDayEventComponent, [{
        type: Component,
        args: [{
                selector: 'mwl-calendar-all-day-event',
                template: "\n    <ng-template\n      #defaultTemplate\n      let-event=\"event\"\n      let-eventClicked=\"eventClicked\">\n      <div\n        class=\"cal-all-day-event\"\n        [style.backgroundColor]=\"event.color?.secondary\"\n        [style.borderColor]=\"event.color?.primary\">\n        <mwl-calendar-event-actions [event]=\"event\"></mwl-calendar-event-actions>\n        &ngsp;\n        <mwl-calendar-event-title\n          [event]=\"event\"\n          [customTemplate]=\"eventTitleTemplate\"\n          view=\"day\"\n          (mwlClick)=\"eventClicked.emit()\">\n        </mwl-calendar-event-title>\n      </div>\n    </ng-template>\n    <ng-template\n      [ngTemplateOutlet]=\"customTemplate || defaultTemplate\"\n      [ngTemplateOutletContext]=\"{\n        event: event,\n        eventClicked: eventClicked\n      }\">\n    </ng-template>\n  "
            }]
    }], function () { return []; }, { eventClicked: [{
            type: Output
        }], event: [{
            type: Input
        }], customTemplate: [{
            type: Input
        }], eventTitleTemplate: [{
            type: Input
        }] }); })();
    return CalendarAllDayEventComponent;
}());
export { CalendarAllDayEventComponent };
function CalendarAllDayEventComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    CalendarAllDayEventComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    CalendarAllDayEventComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    CalendarAllDayEventComponent.propDecorators;
    /** @type {?} */
    CalendarAllDayEventComponent.prototype.event;
    /** @type {?} */
    CalendarAllDayEventComponent.prototype.customTemplate;
    /** @type {?} */
    CalendarAllDayEventComponent.prototype.eventTitleTemplate;
    /** @type {?} */
    CalendarAllDayEventComponent.prototype.eventClicked;
}
