// ng-uikit-pro-standard/esm2015/lib/pro/timepicker/timepicker.directive.js
import { Directive, Input, forwardRef, ElementRef, EventEmitter, HostListener, } from '@angular/core';
import { MdbTimePickerComponent } from './timepicker.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
export const MDB_TIMEPICKER_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    // tslint:disable-next-line: no-use-before-declare
    useExisting: forwardRef(() => MdbTimePickerDirective),
    multi: true,
};
export class MdbTimePickerDirective {
    constructor(el) {
        this.el = el;
        this._valueChange = new EventEmitter();
        this.onChange = () => { };
        this.onTouched = () => { };
    }
    set value(value) {
        this._value = value;
        this._valueChange.emit(this._value);
        this.el.nativeElement.value = value;
    }
    get value() {
        return this._value;
    }
    handleInput(event) {
        this.onChange(event.target.value);
        this._valueChange.emit(event.target.value);
    }
    ngOnInit() {
        this.mdbTimePicker.setInput(this);
        this.mdbTimePicker._selectionChange$.subscribe(selectedValue => {
            this.value = selectedValue;
            this.onChange(selectedValue);
            this.onTouched();
        });
    }
    writeValue(value) {
        if (value || value === '') {
            this.el.nativeElement.value = value;
            this.mdbTimePicker._selectionChange$.next(this._value);
        }
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
}
MdbTimePickerDirective.ɵfac = function MdbTimePickerDirective_Factory(t) { return new (t || MdbTimePickerDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
MdbTimePickerDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: MdbTimePickerDirective, selectors: [["", "mdbTimePicker", ""]], hostBindings: function MdbTimePickerDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("blur", function MdbTimePickerDirective_blur_HostBindingHandler($event) { return ctx.onTouched($event); })("change", function MdbTimePickerDirective_change_HostBindingHandler($event) { return ctx.onChange($event.target.value); })("input", function MdbTimePickerDirective_input_HostBindingHandler($event) { return ctx.handleInput($event); });
    } }, inputs: { value: "value", mdbTimePicker: "mdbTimePicker" }, features: [ɵngcc0.ɵɵProvidersFeature([MDB_TIMEPICKER_VALUE_ACCESSOR])] });
MdbTimePickerDirective.ctorParameters = () => [
    { type: ElementRef }
];
MdbTimePickerDirective.propDecorators = {
    mdbTimePicker: [{ type: Input }],
    value: [{ type: Input }],
    handleInput: [{ type: HostListener, args: ['input', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MdbTimePickerDirective, [{
        type: Directive,
        args: [{
                selector: '[mdbTimePicker]',
                // tslint:disable-next-line: no-host-metadata-property
                host: { '(blur)': 'onTouched($event)', '(change)': 'onChange($event.target.value)' },
                providers: [MDB_TIMEPICKER_VALUE_ACCESSOR]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { value: [{
            type: Input
        }], handleInput: [{
            type: HostListener,
            args: ['input', ['$event']]
        }], mdbTimePicker: [{
            type: Input
        }] }); })();
