// ng-uikit-pro-standard/esm2015/lib/pro/animated-cards/animated-cards.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CardRevealComponent } from './card-reveal.component';
import { CardRotatingComponent } from './card-rotating.component';
import * as ɵngcc0 from '@angular/core';
export class AnimatedCardsModule {
    static forRoot() {
        return { ngModule: AnimatedCardsModule, providers: [] };
    }
}
AnimatedCardsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AnimatedCardsModule });
AnimatedCardsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AnimatedCardsModule_Factory(t) { return new (t || AnimatedCardsModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AnimatedCardsModule, { declarations: function () { return [CardRevealComponent, CardRotatingComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [CardRevealComponent, CardRotatingComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AnimatedCardsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [CardRevealComponent, CardRotatingComponent],
                exports: [CardRevealComponent, CardRotatingComponent]
            }]
    }], null, null); })();
